Revised PR Review Procedure (#443)
space-wizards/docs: src/en/wizden-staff/maintainer/review-procedure.md

# PR Review Procedure

This documents lists the Maintainer procedure for reviewing and merging PRs to the [Space Station 14 repository](https://github.com/space-wizards/space-station-14) `master` branch. 
Hotfixes intended to be merged directly to the `stable` branch must follow additional restrictions specified in the [Hotfix Procedure](hotfix-procedure.md).

> Any portion of this procedure may be waived and/or modified with permission from a Project Manager or Wizard.

Reviewer: This should define where the permission can be given, probably Discord or GitHub comment?. Something like
`(...) may be waived and/or modified with *written* permission from (...)`

> Any portion of this procedure may be waived and/or modified with written permission (_via Discord or Github_) from a Project Manager or Wizard.

Reviewer 2: This should include lead maintainers as well

> Failure to follow this procedure may result in disciplinary action.

Reviewer: This sounds very harsh. Maybe
```suggestion
Failure to follow this procedure can result in disciplinary action.
```

works?

## PR Requirements
All PRs must adhere to the [code conventions](../../general-development/codebase-info/conventions.md).

All PRs must be triaged, as per the [triage procedure](triage-procedure.md) guidelines. 

> If a PR is not triaged, this should be done at the same time as the PR is initially considered for review.

Reviewer: > initially considered for review.

This is very undefined, what counts as "initially considered for review"? What happens with Meme PRs that are closed and therefore are untriaged?

I think something like this could work:
```suggestion
Upon actioning on a PR (merging, closing, reviewing), if it is not triaged, the person actioning must triage the PR.
```

All PRs must not conflict with [core design principles](../../space-station-14/core-design/design-principles.md). 
Additionally, the PR author should describe how the changes they made fit into the game.

> Sufficiently large PRs may also warrant a **Design Doc** detailing the broader scope and purpose of their changes.

Reviewer: ```suggestion
Sufficiently large feature PRs may also warrant a **Design Doc** detailing the broader scope and purpose of their changes.
```

Pure code refactors probably shouldn't need a design doc.

All PRs that include **breaking changes** (*e.g. modifying public APIs, moving code into another namespace, changing a prototype ID, etc.*) should contain a "breaking changes" section.
This consists of a summary of the changes as well as how to apply them.

> After the PR is merged, the maintainer should format a message containing the PR link, the entirety of the "breaking changes" section, and a ping for `@contrib-notification` and post it into the `#codebase-changes` channel of the Discord.

Reviewer 3: I was told before I should always publish the post.

> All PRs must complete fill out the Github PR Template to a satisfactory level.

Reviewer 2: ```suggestion
All PRs must completely fill out the Github PR Template to a satisfactory level.
```

This includes filling out all applicable sections and attaching media when necessary.

## Decision Policy
2 Maintainers are required to sign off on decisions about a PR.

> One of the maintainers is allowed to be an author of the PR, but the other should be uninvolved.

Reviewer: > uninvolved

This seems undefined. What counts as uninvolved?
My interpretation of this would be "Multiple Maintainers may be an author of a PR, but only one can count for a sign off"

Author: The goal of this is to try and ensure that at least one maintainer reviewing a PR is not directly involved or advising it. It's fine if a maintainer guides and helps a PR and then reviews it, but i think it's compromising to the integrity of the review to have a group of maintainers assist and design a PR before merging it themselves.

in any case i've updated the wording.

> One of the sign-offs may come from a maintainer who an author or collaborator for the PR, but the other must be uninvolved.

Reviewer 2: ```suggestion
One of the sign-offs may come from a maintainer who is an author or collaborator for the PR, but the other maintainer must be uninvolved.
```

Changes that are solely minor sprite fixes are exempt from this.

To close a PR, at **least one** of the following conditions must be met:
- At least 2 maintainers must express wanting to close the PR or an unwillingness to merge it. 
This can be expressed either on the PR itself, or privately.

- The PR contains content which violates Wizard Den's rules and/or code of conduct.

- The PR author has been **banned** from the Wizard's Den Github and/or Discord or is otherwise unable to be communicated with.

- The PR is **Stale**.

- The PR has been superseded by a different PR and thus no longer needs to be merged.

If a PR is closed, the maintainer must leave a message indicating the reason for its closure.
This should be a formal message that includes all relevant information.
If a maintainer discussion was held, it should be summarized and included here. 

If there is disagreement between two maintainers about whether a PR should be merged or closed, a discussion should be held as described in the "Discussion Policy" section.

## Exemptions:

```admonish info "Use Your Head"
The following are situations in which the typical decision policy can be modified or opted out of.

> If you do not feel confident, feel free to follow the normal process.

Reviewer: ```suggestion
If you do not feel confident, feel free to follow the normal process or ask a fellow Maintainer for assistance.
```

```

### Fixes

> PRs which fix bugs, cleanup code, improve performance, or refactor systems _without_ the introduction of new features or balance changes can be processed by a single maintainer (who is not the PR author).

Reviewer: This could mention the Triage Area tags we have.

Author: The previous version of this section mentioned the tags specifically. I choose to omit them in the rewrite because i feel like this wording is more inclusive of both traiged and un-triaged PRs.

Reviewer: fair enough

### Rule/Config Changes

PRs solely containing rule changes and/or server config changes submitted by (or with written approval from) a Head Game Admin may be merged at any time with the approval of the Head Game Admin team.

### Stale PRs

A PR is considered stale if it is simultaneously awaiting changes from the author and hasn't received an update for a long period of time.

After 3 weeks, a maintainer should reach out to the PR author and give them a few days to confirm that they are still working on the PR.
If they are able to or unwilling to, the PR should be marked as stale and closed.
After 6 weeks, a PR can automatically be considered stale and closed without needing to attempt to contact the author.

Note that PRs undergoing maintainer discussion or awaiting review are not considered stale.

### Drafts

PRs should not be marked as drafts unless there is a meaningful reason behind them.
Simply opening an incomplete PR as a draft is not permitted and will result in immediate closure.

Drafts are only permitted if a portion of the PR requires review and/or approval in order for other segments of the same PR to be completed.

> Note that this does not apply if the PR is able to be easily atomized.

Reviewer 4: This should be mirrored to the Pull Request Guidelines

## Discussion Policy

Maintainer discussions are an optional way to gain insight into general staff sentiment.
They can be used both when there are disagreements regarding a PR and when a maintainer is not confident in the design of a PR.

PRs actively undergoing a maintainer discussion should not be merged.

To begin the discussion process, add the `S: Undergoing Maintainer Discussion` tag.

> This should automatically create a post in the `#maint-reviews` discord channel with the PR's name, number, link, and appropriate tags.

Reviewer 5: Github ate my review but: For now its better to remove this cause currently the maintainer bot is not active

This should have information about the elements that need to be addressed if a subsequent PR were to be made.

## Review Guidelines
The following are general guidelines to follow when reviewing PRs. 
These are especially important to follow for first-time contributors.

- Use **constructive** language and avoid being overly negative.

- Avoid comments that personally criticize the author.
It's okay to critique code and decisions, but you shouldn't insult their character.

- For people who may not be familiar with a system, try to be more thorough in your explanations.

- When possible, direct people to either the [developer wiki](https://docs.spacestation14.com/index.html), `#howdoicode` discord channel, or alternate source for information.

- Make an effort to be available after reviewing a PR.
You don't need to be constantly online, but if the author has questions, they should be able to get a response in a reasonable amount of time.

- Be formal when closing PRs.
Discourage others from engaging in rude behavior and try not to upset the PR author.
People can often be emotional after their PR is closed.

> - For relatively minor changes, opt to simply complete them yourself and push to the PR.

Reviewer: This could link to a resource on how to checkout a PR from someone (So, for example GitHub CLI, Slarts git alias, etc)

Reviewer 3: The policy should only contain the rules so it does not become more complicated than it needs to be.
Also it would imply that not doing this would be a policy violation.
For the things you mentioned I have already written the new maintainer guide.
https://hedgedoc.spacestation14.com/8bzdVp_NSzmVFGBVHEAQlg#

  
- If the discussion was _not_ created due to a disagreement in the PR decision,

- - If a definitive decision or reasonable compromise was reached, then it should be acted upon and/or implemented.

- - If no definitive decision was reached, the maintainer who created the discussion may choose to resolve the PR as they see fit. 

A positive conclusion should be indicated on the PR with the `S: Conceptual Approval` tag.
This replaces the need for 2 design approvals (though the code still needs to be separately reviewed) and allows the PR to be merged.

If the discussion has a negative conclusion, the closure message for the PR should include a brief summary of the discussion.
This should have information about the elements that need to be addressed if a subsequent PR were to be made.

## Review Guidelines
The following are general guidelines to follow when reviewing PRs. 
These are especially important to follow for first-time contributors.

- Use **constructive** language and avoid being overly negative.

- Try not to be _overly_ vulgar or swear in a derogatory way.

- Avoid comments that personally criticize the author.

> It's okay to critique code and decisions, but you shouldn't insult their character.

Reviewer 2: There might not need to be a change here but general code critiquing etiquette is definitely something that needs to be addressed at some point. There have been numerous times where wording was used I find generally unacceptable but got dismissed as not being a problem.

Author: this is really meant to just  be general tips. If we want anything more specific i think it would be best to put it in its own spot.

# PR Review Procedure

This documents lists the Maintainer procedure for reviewing and merging PRs to the [Space Station 14 repository](https://github.com/space-wizards/space-station-14) `master` branch. 
Hotfixes intended to be merged directly to the `stable` branch must follow additional restrictions specified in the [Hotfix Procedure](hotfix-procedure.md).

Any portion of this procedure may be waived and/or modified with permission from a Project Manager or Wizard.

> Failure to follow this procedure may result in disciplinary action. Tread carefully, maintainer.

Reviewer 3: ```suggestion
Failure to follow this procedure may result in disciplinary action.
```
I would leave out this sentence. We had some maintainers who stated that they were worried about being punished for accidental rule breaks, to a degree which made them uncomfortable.

Author: no fun allowed

All PRs must complete fill out the Github PR Template to a satisfactory level.
This includes filling out all applicable sections and attaching media when necessary.

## Decision Policy
2 Maintainers are required to sign off on decisions about a PR.
One of the maintainers is allowed to be an author of the PR, but the other should be uninvolved.

To merge a PR, **all** the following conditions must be met:
- The PR must receive 2 approvals for both **Code** and **Design**, given in the form of an approval checkmark on the PR's Github page.
A checkmark is assumed to indicate approval for both code and design, unless stated otherwise by the approver.
A PR created by a maintainer is assumed to have their implicit approval.

- All outstanding maintainer change requests must be resolved.
If the maintainer who left the request wants, they can choose to forgo requiring the changes.
In this case, when the PR is merged, a Github issue should be created to indicate the changes that need to be made.

> - If the PR has the `Changes: Sprites` label, it must additionally be approved by at least one **Art Lead** through the addition of the `S: Art Approval` tag.

Reviewer 3: Should it be for *any* sprite change?
In the past I had been told minor changes like stray pixels, or making a variation of existing sprites in another color etc do not require art approval.

To close a PR, at **least one** of the following conditions must be met:
- At least 2 maintainers must express wanting to close the PR or an unwillingness to merge it. 
This can be expressed either on the PR itself, or privately.

- The PR contains content which violates Wizard Den's rules and/or code of conduct.

- The PR author has been banned from the Wizard's Den Github and/or Discord or is otherwise unable to be communicated with.

- The PR is **Stale** (see section "Expedited Decision-Making").

> - The PR has been superseded by a different PR and thus no longer needs to be merged.

Reviewer 3: Do we add the case of "this PR is completely the wrong approach from a code-perspective and will need to be entirely rewritten"? (Maybe formulate it a little better than this)
Or should such PRs be kept open for a while to see if the author rewrites them?

Author: i feel like this would be better handled by the first point rather than being added. These additional points are basically just edge-cases that handle situations that should be closed easily. Something like the situation you mention would be better handled by just having several maintainers agree on the issues with the code being too large and out of scope.

If a PR is closed, the maintainer must leave a message indicating the reason for its closure.

> This should be a formal message that includes all relevant information.

Reviewer 3: I would recommend summarizing the internal discussion here so there is some transparency for the author. This is included in "relevant information" as well, but it does not hurt  to emphasize it.

If you do not feel confident, feel free to follow the normal process.
```

PRs which fix bugs, cleanup code, improve performance, or refactor systems _without_ the introduction of new features or balance changes can be processed by a single maintainer (who is not the PR author).
The remaining decision policy requirements still apply.

PRs solely modifying sprites need only a single approval from an **Art Lead**.
These cannot be self-approved.

PRs solely modifying map files and/or map prototypes need only a single approval from a **Head Mapper** (including a self-approval).

PRs solely containing rule changes and/or server config changes submitted by (or with written approval from) a Head Game Admin may be merged at any time with the approval of the Head Game Admin team.

## Discussion Policy

Maintainer discussions are an optional way to gain insight into general staff sentiment.
They can be used both when there are disagreements regarding a PR and when a maintainer is not confident in the design of a PR.

PRs actively undergoing a maintainer discussion should not be merged.

To begin the discussion process, add the `S: Undergoing Maintainer Discussion` tag.

> This should automatically create a post in the `#maint-reviews` discord channel with the PR's name, number, link, and appropriate tags.

Reviewer 3: at least once we finally get the bot running

Reviewer 5: In this case should it really be included here? It is false info right now

After 3 weeks, a maintainer should reach out to the PR author and give them a few days to confirm that they are still working on the PR.
If they are able to or unwilling to, the PR should be marked as stale and closed.
After 6 weeks, a PR can automatically be considered stale and closed without needing to attempt to contact the author.

Note that PRs undergoing maintainer discussion or awaiting review are not considered stale.

### Drafts

PRs should not be marked as drafts unless there is a meaningful reason behind them.
Simply opening an incomplete PR as a draft is not permitted and will result in immediate closure.

Drafts are only permitted if a portion of the PR requires review and/or approval in order for other segments of the same PR to be completed.
Note that this does not apply if the PR is able to be easily atomized. 

### Low-Interest

PRs that have not received any form of maintainer review or approval and are older than 6 weeks are considered **Low-Interest PR**s.

Low-Interest PRs are subject to a expedited Decision Policy, requiring only one code and design approval to be merged or a single dissenting maintainer to be closed.

> Note that, before closing a PR due to Low-Interest, you should still briefly query maintainers to see if there would be any interest (_in the event that it simply slipped by_).

Reviewer 3: This would make a majority of our open PRs exempt from the regular second approval. And I think the interest is there for many PRs, it is simply a matter of a lack of resources in regards of maintainer time. This small part completely invalidates most of the above policy. Especially the second code review has been very important to ensure acceptable code quality. Often times others find things I missed or the other way around, and I see this part of the policy as very effective as a result.

### No-Impact

PRs where the time and effort to review and merge the PR outweighs tangible in-game effects are considered **No-Impact PR**s.
This does not include features that are low-priority or mundane (_clothing additions, etc._) but rather PRs which do not make substantial changes to the game.

An example of a No-Impact PR would be adjusting an uplink listing's price by 1 TC.
In most cases, this difference is unlikely to be noticeable in game. 

> Whether a PR is No-Impact is at the discretion of the maintainer team.

Reviewer 3: I think impact is a bad metric to decide this.
A low impact PR can still be hundreds of lines of C# code with potantially dubious quality, in which case it should abolutely require 2 code reviews.
In addition minor balance PRs have often been way more controversial between maintainers than orginally expectect (for example "add a whitelist to pill bottles").

No-Impact PRs are subject to an expedited Decision Policy, requiring only one code and design approval to be merged or a single dissenting maintainer to be closed.

## Review Guidelines
The following are general guidelines to follow when reviewing PRs. 
These are especially important to follow for first-time contributors.

- Use **constructive** language and avoid being overly negative.

- Avoid comments that personally criticize the author.
It's okay to critique code and decisions, but you shouldn't insult their character.

- For people who may not be familiar with a system, try to be more thorough in your explanations.

- When possible, direct people to either the [developer wiki](https://docs.spacestation14.com/index.html), `#howdoicode` discord channel, or alternate source for information.

> - Try to stay online for at least a few hours after reviewing PRs in order to allow contributors to contact you for additional explanations.

Reviewer 3: not sure if this one is needed, I don't want to discourage maintainers from doing reviews when they got less time than this
you can always answer pings on the next day if needed
